toto/in_toto_record.py

```diff
@@ -0,0 +1,221 @@
+  """
+  <Purpose>
+    Starts creating link metadata for a multi-part in-toto step. I.e.
+    loads link signing private key from disk, records passed materials, creates
+    link meta data object from it, signs it with loaded key and stores it to
+    disk as ".<step_name>.link-unfinished".
+
+  <Arguments>
+    step_name:
+            A unique name to relate link metadata with a step defined in the
+            layout.
+    key_path:
+            Path to private key to sign link metadata (PEM).
+    material_list:
+            List of file or directory paths that should be recorded as
+            materials.
+
+  <Exceptions>
+    None.
+
+  <Side Effects>
+    Loads private key from disk
+    Creates a file with  name ".<step_name>.link-unfinished"
+    Calls sys.exit(1) if an exception is raised
+
+  <Returns>
+    Link object (unfinished)
+
+  """
+
+  unfinished_fn = "." + str(step_name) + ".link-unfinished"
+
+  try:
+    log.doing("Start recording '{0}'...".format(step_name))
+    key = toto.util.prompt_import_rsa_key_from_file(key_path)
+
+    log.doing("record materials...")
+    materials_dict = runlib.record_artifacts_as_dict(material_list)
+
+    log.doing("create preliminary link metadata...")
+    link = runlib.create_link_metadata(step_name, materials_dict)
+
+    log.doing("sign metadata...")
+    link.sign(key)
+
+    log.doing("store metadata to '{0}'...".format(unfinished_fn))
+    link.dump(unfinished_fn)
+
+    return link
+
+  except Exception, e:
+    log.error("in start record - %s" % e)
+    sys.exit(1)
+
+def in_toto_record_stop(step_name, key_path, product_list):
+  """
+  <Purpose>
+    Finishes creating link metadata for a multi-part in-toto step.
+    Loads signing key and unfinished link metadata file from disk, verifies
+    that the file was signed with the key, records products, updates unfinished
+    Link object (products and signature), removes unfinished link file from and
+    stores new link file to disk.
+
+  <Arguments>
+    step_name:
+            A unique name to relate link metadata with a step defined in the
+            layout.
+    key_path:
+            Path to private key to verify unfinished link metadata and
+            sign finished link metadata (PEM).
+    product_list:
+            List of file or directory paths that should be recorded as products.
+
+  <Exceptions>
+    Exception if signature fails
+    IOException if file ".<step_name>.link-unfinished" can't be read
+
+  <Side Effects>
+    Loads private key from disk
+    Writes link file to disk "<step_name>.link"
+    Remvoes unfinished link file ".<step_name>.link-unfinished" from disk
+    Calls sys.exit(1) if an exception is raised
+
+  <Returns>
+    Link object
+
+  """
+
+  unfinished_fn = "." + str(step_name) + ".link-unfinished"
+
+  try:
+    log.doing("load link signing key...")
+    key = toto.util.prompt_import_rsa_key_from_file(key_path)
+    keydict = {key["keyid"] : key}
```

Comment:
does this make the keyid be repeated inside the value? maybe we don't want to have this.